File: mobile/ios/Air/SubModules/WalletCore/Deeplink/DeeplinkParser.swift
import Foundation
import WalletContext

public extension Deeplink {
    init?(url: URL) {
        let deeplink: Deeplink? = if isWalletConnectPayUrl(url) {
            parseWalletConnectUrl(url)
        } else {
            switch url.scheme?.lowercased() {
            case "ton":
                parseTonInvoiceUrl(url)
            case "tc":
                parseTonConnectUrl(url)
            case let scheme? where scheme == TONCONNECT_PROTOCOL_SCHEME:
                parseTonConnectUrl(url)
            case "wc":
                parseWalletConnectUrl(url)
            case let scheme? where compatibleWalletConnectSelfProtocolSchemes.contains(scheme) && url.host?.lowercased() == "wc":
                parseWalletConnectWrapperUrl(url)
            case let scheme? where compatibleWalletConnectWrapperProtocolSchemes.contains(scheme):
                parseWalletConnectWrapperUrl(url)
            case let scheme? where compatibleSelfProtocolSchemes.contains(scheme):
                parseMtwUrl(url)
            case "https", "http":
                switch url.host?.lowercased() {
                case let host? where isWalletConnectUniversalUrl(host: host, path: url.path):
                    parseWalletConnectWrapperUrl(url)
                case let host? where isTonConnectUniversalUrl(host: host, path: url.path):
                    parseTonConnectUrl(url)
                case "walletconnect.com":
                    if url.path == "/wc" {
                        parseWalletConnectWrapperUrl(url) ?? parseWalletConnectUrl(url)
                    } else {
                        nil
                    }
                case let host? where compatibleSelfUniversalHosts.contains(host):
                    parseMtwUrl(url)
                default:
                    nil
                }
            default:
                nil
            }
        }
        if let deeplink {
            self = deeplink
        } else {
            return nil
        }
    }
}

private let gramLegacySelfProtocolScheme = "mtw"
private let gramLegacySelfUniversalHosts: Set<String> = ["my.tt", "go.mytonwallet.org"]
private let compatibleWalletConnectSelfProtocolSchemes: Set<String> = ["mtw", "twalletgram"]
private let compatibleWalletConnectWrapperProtocolSchemes: Set<String> = ["mw", "mywallet-wc", "gramwallet-wc"]
private let compatibleWalletConnectUniversalHosts: Set<String> = [
    "connect.mywallet.io",
    "connect.mytonwallet.org",
    "connect.gramwallet.io",
]

private var compatibleSelfProtocolSchemes: Set<String> {
    var schemes: Set<String> = [SELF_PROTOCOL_SCHEME]
    if IS_TWALLETGRAM_WALLET {
        schemes.insert(gramLegacySelfProtocolScheme)
    }
    return schemes
}

private var compatibleSelfUniversalHosts: Set<String> {
    var hosts = SELF_UNIVERSAL_URL_HOSTS
    if IS_TWALLETGRAM_WALLET {
        hosts.formUnion(gramLegacySelfUniversalHosts)
    }
    return hosts
}

private func normalizeSelfProtocolUrl(_ url: URL) -> URL? {
    guard let scheme = url.scheme?.lowercased(),
          scheme != SELF_PROTOCOL_SCHEME,
          compatibleSelfProtocolSchemes.contains(scheme),
          var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
        return nil
    }
    components.scheme = SELF_PROTOCOL_SCHEME
    return components.url
}

private func normalizeSelfUniversalUrl(_ url: URL) -> URL? {
    guard let scheme = url.scheme?.lowercased(),
          scheme == "https" || scheme == "http",
          let host = url.host?.lowercased(),
          compatibleSelfUniversalHosts.contains(host),
          let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
        return nil
    }

    var path = components.percentEncodedPath
    if path.hasPrefix("/") {
        path.removeFirst()
    }
    var urlString = SELF_PROTOCOL + path
    if let query = components.percentEncodedQuery {
        urlString += "?\(query)"
    }
    if let fragment = components.percentEncodedFragment {
        urlString += "#\(fragment)"
    }
    return URL(string: urlString)
}

private func isWalletConnectUniversalUrl(host: String, path: String) -> Bool {
    guard compatibleWalletConnectUniversalHosts.contains(host.lowercased()) else {
        return false
    }
    return isWalletConnectWrapperPath(path)
}

private func isWalletConnectWrapperPath(_ path: String) -> Bool {
    let normalizedPath = path.trimmingCharacters(in: CharacterSet(charactersIn: "/")).lowercased()
    return normalizedPath == "wc" || normalizedPath == "wc/wc"
}

private func isTonConnectUniversalUrl(host: String, path: String) -> Bool {
    guard let url = URL(string: TONCONNECT_UNIVERSAL_URL),
          let universalHost = url.host?.lowercased() else {
        return false
    }
    guard host.lowercased() == universalHost else {
        return false
    }
    let universalPath = url.path
    if universalPath.isEmpty {
        return true
    }
    return path == universalPath
}

private func isWalletConnectPayUrl(_ url: URL) -> Bool {
    let absoluteString = url.absoluteString
    if absoluteString.hasPrefix("pay_") {
        return true
    }
    if absoluteString.contains("?pay=") || absoluteString.contains("&pay=") {
        return true
    }
    if let host = url.host?.lowercased(), isWalletConnectPayHost(host) {
        return true
    }
    return false
}

private func isWalletConnectPayHost(_ host: String) -> Bool {
    let host = host.lowercased()
    return host == "pay.walletconnect.com"
        || host == "pay.walletconnect.org"
        || host.hasSuffix(".pay.walletconnect.com")
        || host.hasSuffix(".pay.walletconnect.org")
}

private func parseTonConnectUrl(_ url: URL) -> Deeplink {
    Deeplink.tonConnect2(requestLink: url.absoluteString)
}

private func parseWalletConnectUrl(_ url: URL) -> Deeplink {
    Deeplink.walletConnect(requestLink: url.absoluteString)
}

private func parseWalletConnectWrapperUrl(_ url: URL) -> Deeplink? {
    guard let requestLink = walletConnectRequestLink(from: url) else {
        return nil
    }
    return Deeplink.walletConnect(requestLink: requestLink)
}

private func walletConnectRequestLink(from url: URL) -> String? {
    guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
        return nil
    }
    if let percentEncodedQuery = components.percentEncodedQuery,
       let value = extractUriValue(from: percentEncodedQuery) {
        let requestLink = value.removingPercentEncoding ?? value
        return requestLink.lowercased().hasPrefix("wc:") ? requestLink : nil
    }
    if let requestLink = components.queryItems?.first(where: { $0.name == "uri" })?.value,
       requestLink.lowercased().hasPrefix("wc:") {
        return requestLink
    }
    return nil
}

private func extractUriValue(from percentEncodedQuery: String) -> String? {
    if percentEncodedQuery.hasPrefix("uri=") {
        return String(percentEncodedQuery.dropFirst("uri=".count))
    }
    if let range = percentEncodedQuery.range(of: "&uri=") {
        return String(percentEncodedQuery[range.upperBound...])
    }
    return nil
}

private func parseTonInvoiceUrl(_ url: URL) -> Deeplink? {
    guard let parsedWalletURL = parseTonTransferUrl(url) else {
        return nil
    }
    return Deeplink.invoice(
        address: parsedWalletURL.address,
        amount: parsedWalletURL.amount,
        comment: parsedWalletURL.comment,
        binaryPayload: parsedWalletURL.bin,
        token: parsedWalletURL.token ?? TONCOIN_SLUG,
        jetton: parsedWalletURL.jetton,
        stateInit: parsedWalletURL.stateInit
    )
}

private func parseMtwUrl(_ url: URL) -> Deeplink? {

    var url = url

    if let normalizedUrl = normalizeSelfProtocolUrl(url) {
        url = normalizedUrl
    }

    if url.scheme == "http", var components = URLComponents(url: url, resolvingAgainstBaseURL: false) {
        components.scheme = "https"
        if let newUrl = components.url {
            url = newUrl
        }
    }

    if let normalizedUrl = normalizeSelfUniversalUrl(url) {
        url = normalizedUrl
    }

    switch url.host {
    case "agent":
        return .agent

    case "wc":
        return parseWalletConnectWrapperUrl(url)
        
    case "swap", "buy-with-crypto":
        var from: String? = nil
        var to: String? = nil
        var amountIn: Double? = nil
        if let query = url.query, let components = URLComponents(string: "/?" + query), let queryItems = components.queryItems {
            for queryItem in queryItems {
                if let value = queryItem.value {
                    if queryItem.name == "amountIn", !value.isEmpty, let amountValue = Double(value) {
                        amountIn = amountValue
                    } else if queryItem.name == "in", !value.isEmpty {
                        from = value
                    } else if queryItem.name == "out", !value.isEmpty {
                        to = value
                    }
                }
            }
        }
        if url.host == "buy-with-crypto" {
            if to == nil, from != "toncoin" {
                to = "toncoin"
            }
            if from == nil {
                from = TRON_USDT_SLUG
            }
        }
        return .swap(from: from, to: to, amountIn: amountIn)

    case "transfer":
        if url.pathComponents.count <= 1 {
            return .transfer
        } else {
            return parseTonInvoiceUrl(url)
        }

    case "send":
        return parseSendUrl(url)

    case "buy-with-card":
        return .buyWithCard
        
    case Deeplink.Sell.urlHost:
        return .sell(.init(url))

    case "stake":
        return .stake

    case "portfolio":
        return .portfolio

    case "settings":
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        if let section = pathComponents.first?.nilIfEmpty {
            guard let settingsSection = AppSettingsSection(rawValue: section) else { return nil }
            return .settings(section: settingsSection)
        }
        return .settings(section: nil)

    case "giveaway":
        let pathname = url.absoluteString
        let regex = try! NSRegularExpression(pattern: "giveaway/([^/]+)")
        var giveawayId: String? = nil
        
        if let match = regex.firstMatch(in: pathname, range: NSRange(pathname.startIndex..., in: pathname)) {
            let giveawayIdRange = match.range(at: 1)
            if let giveawayIdRange = Range(giveawayIdRange, in: pathname) {
                giveawayId = String(pathname[giveawayIdRange])
            }
        }
        let urlString = "https://giveaway.mytonwallet.io/\(giveawayId != nil ? "?giveawayId=\(giveawayId!)" : "")"
        let url = URL(string: urlString)!
        return .url(url: url, title: "Giveaway", injectDappConnect: true)
        
    case "r":
        let pathname = url.absoluteString
        let regex = try! NSRegularExpression(pattern: "r/([^/]+)")
        var r: String? = nil
        
        if let match = regex.firstMatch(in: pathname, range: NSRange(pathname.startIndex..., in: pathname)) {
            let rIdRange = match.range(at: 1)
            if let rRange = Range(rIdRange, in: pathname) {
                r = String(pathname[rRange])
            }
        }
        let urlString = "https://checkin.mytonwallet.org/\(r != nil ? "?r=\(r!)" : "")"
        let url = URL(string: urlString)!
        return .url(url: url, title: "Checkin", injectDappConnect: true)
        
    case "receive":
        return .receive
        
    case "explore":
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        return .explore(siteHost: pathComponents.first?.nilIfEmpty)

    case "token":
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        if pathComponents.count >= 2 {
            guard let chain = ApiChain(rawValue: pathComponents[0]) else { return nil }
            let tokenAddress = pathComponents[1]
            return .tokenAddress(chain: chain, tokenAddress: tokenAddress)
        } else if pathComponents.count == 1 {
            let slug = pathComponents[0]
            return .tokenSlug(slug: slug)
        } else {
            return nil
        }
    
    case "tx":
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        guard pathComponents.count >= 2 else { return nil }
        let chainString = pathComponents[0]
        let rawTxId = pathComponents.dropFirst().joined(separator: "/")
        let txId = rawTxId.removingPercentEncoding ?? rawTxId
        guard let chain = ApiChain(rawValue: chainString), chain.isSupported else { return nil }
        return .transaction(chain: chain, txId: txId)

    case "nft":
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        guard let nftAddress = pathComponents.first?.nilIfEmpty else { return nil }
        return .nftAddress(nftAddress: nftAddress)
        
    case "view":
        var addressOrDomainByChain: [String: String] = [:]
        let queryItems = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
        let network: ApiNetwork = queryItems.contains { $0.name == "testnet" && $0.value?.lowercased() == "true" } ? .testnet : .mainnet
        for queryItem in queryItems {
            if let chain = ApiChain(rawValue: queryItem.name),
               chain.isSupported,
               let addressOrDomain = queryItem.value?.nilIfEmpty,
               chain.isValidAddressOrDomain(addressOrDomain) {
                addressOrDomainByChain[chain.rawValue] = addressOrDomain
            }
        }
        if let evmAddress = queryItems.first(where: { $0.name == ApiChain.viewAccountEvmParam })?.value?.nilIfEmpty,
           ApiChain.ethereum.isValidAddressOrDomain(evmAddress) {
            for chain in ApiChain.evmChains where addressOrDomainByChain[chain.rawValue] == nil {
                addressOrDomainByChain[chain.rawValue] = evmAddress
            }
        }
        return .view(network: network, addressOrDomainByChain: addressOrDomainByChain)

    default:
        return nil
    }
}

private func parseSendUrl(_ url: URL) -> Deeplink? {
    let pathComponents = url.pathComponents.filter { $0 != "/" }
    guard let target = pathComponents.first else { return nil }

    guard let colonIndex = target.firstIndex(of: ":") else { return nil }
    let chainString = String(target[target.startIndex..<colonIndex])
    let address = String(target[target.index(after: colonIndex)...])

    guard let chain = ApiChain(rawValue: chainString), chain.isSupported else { return nil }
    guard chain.isValidAddressOrDomain(address) else { return nil }

    let queryItems = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
    var amount: BigInt?
    var comment: String?
    var binaryPayload: String?
    var tokenSlug: String?
    var stateInit: String?

    for item in queryItems {
        guard let value = item.value, !value.isEmpty else { continue }
        switch item.name {
        case "amount":
            amount = BigInt(value)
        case "text":
            comment = value
        case "bin":
            binaryPayload = value
        case "token":
            tokenSlug = value
        case "init", "stateInit":
            stateInit = value
        default:
            break
        }
    }

    return .send(
        chain: chain,
        address: address,
        amount: amount,
        comment: comment,
        binaryPayload: binaryPayload,
        tokenSlug: tokenSlug,
        stateInit: stateInit
    )
}

public extension Deeplink {
    struct Sell: Sendable {
        public static let urlHost = "offramp"
        
        public let transactionId: String?
        public let baseCurrencyCode: String?
        public let baseCurrencyAmount: String?
        public let depositWalletAddress: String?
        public let depositWalletAddressTag: String?

        public init(_ url: URL) {
            let queryItems = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
            var transactionId: String?
            var baseCurrencyCode: String?
            var baseCurrencyAmount: String?
            var depositWalletAddress: String?
            var depositWalletAddressTag: String?

            for item in queryItems {
                guard let value = item.value, !value.isEmpty else { continue }
                switch item.name {
                case "transactionId": transactionId = value
                case "baseCurrencyCode": baseCurrencyCode = value
                case "baseCurrencyAmount": baseCurrencyAmount = value
                case "depositWalletAddress": depositWalletAddress = value
                case "depositWalletAddressTag": depositWalletAddressTag = value
                default: break
                }
            }
            
            self.transactionId = transactionId
            self.baseCurrencyCode = baseCurrencyCode
            self.baseCurrencyAmount = baseCurrencyAmount
            self.depositWalletAddress = depositWalletAddress
            self.depositWalletAddressTag = depositWalletAddressTag
        }
    }
}

File: mobile/android/air/SubModules/WalletCore/src/main/java/app/twallet/air/walletcore/deeplink/DeeplinkParser.kt
package app.twallet.air.walletcore.deeplink

import android.content.Intent
import android.net.Uri
import app.twallet.air.walletbasecontext.APP_SCHEME
import app.twallet.air.walletbasecontext.APP_TC_SCHEME
import app.twallet.air.walletbasecontext.R as BaseR
import app.twallet.air.walletbasecontext.utils.ApplicationContextHolder
import app.twallet.air.walletbasecontext.utils.decodeUrlOrNull
import app.twallet.air.walletcontext.helpers.AddressHelpers
import app.twallet.air.walletcontext.helpers.DNSHelpers
import app.twallet.air.walletcontext.models.MBlockchainNetwork
import app.twallet.air.walletcore.TRON_USDT_SLUG
import app.twallet.air.walletcore.models.InAppBrowserConfig
import app.twallet.air.walletcore.models.blockchain.MBlockchain

sealed class Deeplink {
    abstract val accountAddress: String?

    data class TonConnect2(override val accountAddress: String?, val requestUri: Uri) : Deeplink() {
        val isConnectRequest: Boolean
            get() {
                return !requestUri.getQueryParameter("r").isNullOrBlank()
            }
    }

    data class Invoice(
        override val accountAddress: String?,
        val address: String,
        val amount: String?,
        val binary: String?,
        val expiry: Long?,
        val init: String?,
        val jetton: String?,
        val comment: String?,
        val token: String?,
        val hasUnsupportedParams: Boolean
    ) : Deeplink()

    data class Send(
        override val accountAddress: String?,
        val chain: String,
        val address: String,
        val amount: String?,
        val comment: String?,
        val binary: String?,
        val tokenSlug: String?,
        val init: String?,
        val expiry: Long?,
        val hasUnsupportedParams: Boolean
    ) : Deeplink()

    data class Swap(
        override val accountAddress: String?,
        val from: String?,
        val to: String?,
        val amountIn: Double?
    ) : Deeplink()

    data class Receive(override val accountAddress: String?) : Deeplink()

    data class Portfolio(override val accountAddress: String?) : Deeplink()
    data class Explore(
        override val accountAddress: String?,
        val targetUri: Uri?
    ) : Deeplink()

    data class Url(
        override val accountAddress: String?,
        val config: InAppBrowserConfig
    ) : Deeplink()

    data class Transaction(
        override val accountAddress: String?,
        val chain: String?,
        val txId: String?,
        val txHash: String?
    ) : Deeplink()

    data class TokenBySlug(override val accountAddress: String?, val slug: String) : Deeplink()
    data class TokenByAddress(
        override val accountAddress: String?,
        val chain: String,
        val address: String
    ) : Deeplink()

    data class ExpiringDns(override val accountAddress: String?, val domainAddress: String) :
        Deeplink()

    data class Settings(override val accountAddress: String?, val page: String?) : Deeplink()
    data class WalletConnect(override val accountAddress: String?, val requestUri: Uri) : Deeplink()
    data class WalletConnectPay(override val accountAddress: String?, val requestUri: Uri) :
        Deeplink()
    data class SwitchToLegacy(override val accountAddress: String?) : Deeplink()
    data class View(
        override val accountAddress: String?,
        val network: MBlockchainNetwork,
        val addressByChain: Map<String, String>
    ) : Deeplink()

    data class Nft(
        override val accountAddress: String?,
        val network: MBlockchainNetwork,
        val nftAddress: String
    ) : Deeplink()
}

interface DeeplinkNavigator {
    fun handle(deeplink: Deeplink)
}

class DeeplinkParser {

    companion object {
        private val WC_WRAPPER_SCHEMES = setOf("twallet-wc")
        private val WC_WRAPPER_UNIVERSAL_HOSTS = setOf(
            "connect.mywallet.io",
            "connect.mytonwallet.org",
        )

        fun parse(intent: Intent): Deeplink? {
            return parse(intent.data)
        }

        fun parse(uri: Uri?): Deeplink? {
            if (uri == null)
                return null
            return when (uri.scheme) {
                "ton" -> handleTonInvoice(uri)
                "tc", APP_TC_SCHEME -> handleTonConnect(uri)
                "wc" -> handleWalletConnect(uri)
                APP_SCHEME -> handleMTW(uri)
                "https" -> handleHttpsDeeplinks(uri)
                in WC_WRAPPER_SCHEMES -> handleWalletConnectWrapper(uri)
                else -> {
                    null
                }
            }
        }

        private fun handleTonConnect(uri: Uri): Deeplink {
            return Deeplink.TonConnect2(accountAddress = null, requestUri = uri)
        }

        private fun handleWalletConnect(uri: Uri): Deeplink {
            return Deeplink.WalletConnect(accountAddress = null, requestUri = uri)
        }

        private fun handleWalletConnectWrapper(uri: Uri): Deeplink? {
            val requestLink = walletConnectRequestLink(uri) ?: return null
            return Deeplink.WalletConnect(
                accountAddress = null,
                requestUri = Uri.parse(requestLink)
            )
        }

        private fun walletConnectRequestLink(uri: Uri): String? {
            val encodedQuery = uri.encodedQuery
            val encodedValue = encodedQuery?.let { extractUriValue(it) }
            if (encodedValue != null) {
                val requestLink = encodedValue.decodeUrlOrNull() ?: encodedValue
                return if (requestLink.lowercase().startsWith("wc:")) requestLink else null
            }
            val requestLink = uri.getQueryParameter("uri") ?: return null
            return if (requestLink.lowercase().startsWith("wc:")) requestLink else null
        }

        private fun extractUriValue(encodedQuery: String): String? {
            if (encodedQuery.startsWith("uri=")) {
                return encodedQuery.removePrefix("uri=")
            }
            val index = encodedQuery.indexOf("&uri=")
            if (index >= 0) {
                return encodedQuery.substring(index + "&uri=".length)
            }
            return null
        }

        private fun isWalletConnectWrapperPath(path: String?): Boolean {
            val normalizedPath = path?.trim('/')?.lowercase() ?: ""
            return normalizedPath == "wc" || normalizedPath == "wc/wc"
        }

        private fun handleTonInvoice(uri: Uri): Deeplink? {
            val parsedWalletURL = parseWalletUrl(uri) ?: return null
            return Deeplink.Invoice(
                accountAddress = null,
                address = parsedWalletURL.address,
                amount = parsedWalletURL.amount,
                binary = parsedWalletURL.binary,
                comment = parsedWalletURL.comment,
                expiry = parsedWalletURL.expiry,
                init = parsedWalletURL.init,
                jetton = parsedWalletURL.jetton,
                token = parsedWalletURL.token,
                hasUnsupportedParams = parsedWalletURL.hasUnsupportedParams,
            )
        }

        private fun handleHttpsDeeplinks(uri: Uri): Deeplink? {
            val universalHosts = setOf("my.tt", "go.mytonwallet.org")
            val tonconnectHosts = setOf("connect.mytonwallet.org")
            val host = uri.host
            when {
                host != null && host.lowercase() in WC_WRAPPER_UNIVERSAL_HOSTS &&
                    isWalletConnectWrapperPath(uri.path) -> {
                    return handleWalletConnectWrapper(uri)
                }

                host != null && host in universalHosts -> {
                    val path = uri.path?.trimStart('/') ?: ""
                    val pathSegments = path.split('/')

                    val mtwUri = Uri.Builder()
                        .scheme(APP_SCHEME)
                        .authority(pathSegments.firstOrNull() ?: "")
                        .apply {
                            if (pathSegments.size > 1) {
                                pathSegments.drop(1).forEach { segment ->
                                    appendPath(segment)
                                }
                            }
                        }
                        .encodedQuery(uri.encodedQuery)
                        .build()
                    return handleMTW(mtwUri)
                }

                host != null && host in tonconnectHosts -> {
                    return handleTonConnect(uri)
                }

                host == "pay.walletconnect.com" || host == "pay.walletconnect.org" -> {
                    return handleWalletConnectPay(uri)
                }

                host == "walletconnect.com" -> {
                    if (uri.path == "/pay" || uri.path?.startsWith("/pay/") == true)
                        return handleWalletConnectPay(uri)
                    if (uri.path == "/wc") handleWalletConnectWrapper(uri) ?: handleWalletConnect(uri)
                }
            }
            return null
        }

        private fun handleWalletConnectPay(uri: Uri): Deeplink {
            return Deeplink.WalletConnectPay(accountAddress = null, requestUri = uri)
        }

        private fun handleMTW(uri: Uri): Deeplink? {
            return when (uri.host) {
                "swap", "buy-with-crypto" -> {
                    var from: String? = null
                    var to: String? = null
                    var amountIn: Double? = null

                    uri.query?.let { query ->
                        val components = query.decodeUrlOrNull()?.split("&")?.mapNotNull {
                            it.split("=")
                                .let { parts -> if (parts.size == 2) parts[0] to parts[1] else null }
                        }?.toMap() ?: emptyMap()

                        components["amountIn"]?.toDoubleOrNull()?.let { amountIn = it }
                        components["in"]?.let { from = it }
                        components["out"]?.let { to = it }
                    }

                    if (uri.host == "buy-with-crypto") {
                        if (to == null && from != "toncoin") to = "toncoin"
                        if (from == null) from = TRON_USDT_SLUG
                    }

                    Deeplink.Swap(accountAddress = null, from = from, to = to, amountIn = amountIn)
                }

                "wc" -> handleWalletConnectWrapper(uri)
                "transfer" -> handleTonInvoice(uri)
                "send" -> handleSend(uri)
                "receive" -> Deeplink.Receive(accountAddress = null)

                "portfolio" -> Deeplink.Portfolio(accountAddress = null)
                "explore" -> Deeplink.Explore(
                    accountAddress = null,
                    targetUri = uri.extractSubUri()
                )

                "settings" -> {
                    val page = uri.pathSegments.firstOrNull()?.lowercase()
                    Deeplink.Settings(accountAddress = null, page = page)
                }

                "giveaway" -> {
                    val giveawayBase = ApplicationContextHolder.applicationContext
                        .getString(BaseR.string.app_giveaway_url)
                    if (giveawayBase.isEmpty()) return null
                    val giveawayId = extractId(uri.toString(), "giveaway/([^/]+)")
                    val urlString =
                        giveawayBase + if (giveawayId != null) "?giveawayId=$giveawayId" else ""
                    val config = InAppBrowserConfig(
                        url = urlString,
                        title = "Giveaway",
                        injectDappConnect = true
                    )
                    Deeplink.Url(accountAddress = null, config)
                }

                "r" -> {
                    val rId = extractId(uri.toString(), "r/([^/]+)")
                    val urlString =
                        "https://checkin.mytonwallet.org/" + if (rId != null) "?r=$rId" else ""
                    val config = InAppBrowserConfig(
                        url = urlString,
                        title = "Checkin",
                        injectDappConnect = true
                    )
                    Deeplink.Url(accountAddress = null, config)
                }

                "classic" -> Deeplink.SwitchToLegacy(null)

                "token" -> {
                    val pathParts = uri.pathSegments

                    if (pathParts.size > 1) {
                        val chain = pathParts[0]
                        val tokenAddress = pathParts[1]
                        return Deeplink.TokenByAddress(null, chain, tokenAddress)
                    } else {
                        pathParts.firstOrNull()?.let { tokenSlug ->
                            return Deeplink.TokenBySlug(null, tokenSlug)
                        }
                    }
                }

                "tx" -> {
                    val pathParts = uri.pathSegments

                    if (pathParts.size > 1) {
                        val chain = pathParts[0]
                        val rawTxId = pathParts.drop(1).joinToString("/")
                        val txId = rawTxId.decodeUrlOrNull() ?: rawTxId
                        return Deeplink.Transaction(
                            accountAddress = null,
                            chain = chain,
                            txId = txId,
                            txHash = null,
                        )
                    } else {
                        return null
                    }
                }

                "view" -> {
                    val addressByChain = mutableMapOf<String, String>()
                    MBlockchain.supportedChains.forEach { blockchain ->
                        val address = uri.getQueryParameter(blockchain.name)
                        if (!address.isNullOrBlank()) {
                            if (blockchain.isValidAddress(address) || blockchain.isValidDNS(address)) {
                                addressByChain[blockchain.name] = address
                            }
                        }
                    }
                    val evmAddress = uri.getQueryParameter(MBlockchain.VIEW_ACCOUNT_EVM_PARAM)
                    if (!evmAddress.isNullOrBlank() && MBlockchain.ethereum.isValidAddress(
                            evmAddress
                        )
                    ) {
                        MBlockchain.evmChains.forEach { chain ->
                            if (!addressByChain.containsKey(chain.name)) {
                                addressByChain[chain.name] = evmAddress
                            }
                        }
                    }
                    val network =
                        if (uri.getQueryParameter("testnet") == "true") MBlockchainNetwork.TESTNET else MBlockchainNetwork.MAINNET
                    return Deeplink.View(
                        accountAddress = null,
                        network = network,
                        addressByChain = addressByChain
                    )
                }

                "nft" -> {
                    val nftAddress =
                        uri.pathSegments.firstOrNull()?.takeIf { it.isNotBlank() } ?: return null
                    val network =
                        if (uri.getQueryParameter("testnet") == "true") MBlockchainNetwork.TESTNET else MBlockchainNetwork.MAINNET
                    return Deeplink.Nft(
                        accountAddress = null,
                        network = network,
                        nftAddress = nftAddress
                    )
                }

                else -> {
                    return null
                }
            }
        }

        private fun handleSend(uri: Uri): Deeplink? {
            // Format: twallet://send/{chain}:{address}?amount=...&token=...&text=...
            val target = uri.pathSegments.firstOrNull() ?: return null
            val colonIndex = target.indexOf(':')
            if (colonIndex == -1) return null

            val chain = target.substring(0, colonIndex)
            val address = target.substring(colonIndex + 1)

            val blockchain = MBlockchain.valueOfOrNull(chain) ?: return null
            if (!blockchain.isSupported) return null
            if (!blockchain.isValidAddress(address) && !blockchain.isValidDNS(address)) return null

            var amount: String? = null
            var comment: String? = null
            var binary: String? = null
            var tokenSlug: String? = null
            var init: String? = null
            var expiry: Long? = null
            var hasUnsupportedParams = false

            uri.queryParameterNames.forEach { paramName ->
                val value = uri.getQueryParameter(paramName)
                if (!value.isNullOrEmpty()) {
                    when (paramName) {
                        "amount" -> amount = value
                        "text" -> comment = value
                        "bin" -> binary = value
                        "token" -> tokenSlug = value
                        "init", "stateInit" -> init = value
                        "exp" -> try {
                            expiry = value.toLong()
                        } catch (_: NumberFormatException) {
                        }

                        else -> hasUnsupportedParams = true
                    }
                }
            }

            return Deeplink.Send(
                accountAddress = null,
                chain = blockchain.name,
                address = address,
                amount = amount,
                comment = comment,
                binary = binary,
                tokenSlug = tokenSlug,
                init = init,
                expiry = expiry,
                hasUnsupportedParams = hasUnsupportedParams
            )
        }

        private fun extractId(pathname: String, pattern: String): String? {
            val regex = Regex(pattern)
            val match = regex.find(pathname)
            return match?.groups?.get(1)?.value
        }
    }
}

fun Uri.extractSubUri(): Uri? {
    // use only host for now
    val targetHost = pathSegments.firstOrNull() ?: return null
    return Uri.Builder()
        .scheme("https")
        .authority(targetHost)
        .build()
}

fun parseWalletUrl(uri: Uri): ParsedWalletUrl? {
    if ((uri.scheme != "ton" && uri.scheme != APP_SCHEME) || uri.host != "transfer") {
        return null
    }

    val updatedUrl =
        uri.buildUpon().encodedPath(uri.encodedPath).encodedQuery(uri.encodedQuery).build()

    var address: String? = null
    val path = updatedUrl.path?.trim('/') ?: ""
    if (AddressHelpers.isValidAddress(path) || DNSHelpers.isDnsDomain(path)) {
        address = path
    }

    var amount: String? = null
    var binary: String? = null
    var comment: String? = null
    var expiry: Long? = null
    var init: String? = null
    var jetton: String? = null
    var token: String? = null

    var hasUnsupportedParams = false

    updatedUrl.queryParameterNames.forEach { paramName ->
        val value = updatedUrl.getQueryParameter(paramName)
        if (!value.isNullOrEmpty()) {
            when (paramName) {
                "amount" -> amount = value
                "bin" -> binary = value
                "exp" -> try {
                    expiry = value.toLong()
                } catch (e: NumberFormatException) {
                    // Handle invalid amount format
                }

                "init", "stateInit" -> init = value
                "jetton" -> jetton = value
                "text" -> comment = value
                "token" -> token = value
                else -> hasUnsupportedParams = true
            }
        }
    }

    return ParsedWalletUrl(
        address = address ?: "",
        amount = amount,
        binary = binary,
        comment = comment,
        expiry = expiry,
        init = init,
        jetton = jetton,
        token = token,
        hasUnsupportedParams = hasUnsupportedParams
    )
}

data class ParsedWalletUrl(
    val address: String,
    val amount: String?,
    val binary: String?,
    val comment: String?,
    val expiry: Long?,
    val init: String?,
    val jetton: String?,
    val token: String?,
    val hasUnsupportedParams: Boolean
)
